### DISCUSSION TITLE
Syntax for interface delegation

### DISCUSSION THREAD

**Discussion by LukeSawyers** [dotnet/csharplang#5514] [Language Ideas] [answered]
When writing classes trying to follow Composition over Inheritance, things can get a little bit boilerplatey in C#. A feature of other languages such as delphi and kotlin solve this with a bit of syntax and encourage use of the pattern. 

Some discussion on this was had over in the roslyn project that seemed to tail off: https://github.com/dotnet/roslyn/issues/13952

Given interfaces: 

```
public interface IMyInterface
{
    int[] GetThingsDone();
    
    void DoSomething(int i);
}

public interface IMyOtherInterface
{
    string DoSomethingWithResponse();
}
```

And bare portable implementations:

```
public class SimpleInterfaceImpl : IMyInterface
{
    public int[] GetThingsDone() => throw new System.NotImplementedException();

    public void DoSomething(int i) => throw new System.NotImplementedException();
}

public class SimpleOtherInterfaceImpl : IMyOtherInterface
{
    public string DoSomethingWithResponse() => throw new System.NotImplementedException();
}
```

And a more complex class acting as an implementer of both, using the simple implementations

```
public class ComplexImplementer : IMyInterface, IMyOtherInterface
{
    private readonly SimpleInterfaceImpl _impl = new();
    private readonly SimpleOtherInterfaceImpl _otherImpl = new();
    
    public int[] GetThingsDone() => _impl.GetThingsDone();

    public void DoSomething(int i) => _impl.DoSomething(i);

    public string DoSomethingWithResponse() => "Do something different";
}
```
Or a class that acts to wrap a single member of an interface:

```
public class PartialWrapper : IMyInterface
{
    private readonly IMyInterface _wrapped;

    public PartialWrapper(IMyInterface wrapped)
    {
        _wrapped = wrapped;
    }

    public int[] GetThingsDone() => Array.Empty<int>();

    public void DoSomething(int i) => _wrapped.DoSomething(i);
}
```

Examples are short for brevity, in real situations with more code this can get very wordy and laborsome. Refactoring tools can generate this code but it still needs to be maintained and will break if the interface is modified.

Instead implement a language keyword (`by`/`from`/`with`/`using`/`=>`, whatever is most syntactically consistent, I've used `by` below) that specifies delegation to a member, use the `override` keyword to resolve conflicts or specify different behaviour:

```
public class ComplexImplementer : IMyInterface by _impl, IMyOtherInterface by _otherImpl
{
    private readonly SimpleInterfaceImpl _impl = new();
    private readonly SimpleOtherInterfaceImpl _otherImpl = new();

    public override string DoSomethingWithResponse() => "Some different behvaiour";
}
        
public class PartialWrapper : IMyInterface by _wrapped
{
    private readonly IMyInterface _wrapped;

    public PartialWrapper(IMyInterface wrapped)
    {
        _wrapped = wrapped;
    }

    public override int[] GetThingsDone() => Array.Empty<int>();
}
```
 Code is much simpler and clearer, additions to `IMyInterface` and `IMyOtherInterface` do not require the developer to implement passthroughs for `ComplexImplementer` and `PartialWrapper` , Admittedly I do not know how this would play out with explicit interfaces, perhaps `explicit by` could be used to delineate.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/234
